drivers/frequency/adf4382/adf4382.c

```diff
@@ -0,0 +1,1199 @@
+ * @return    		- 0 in case of success or negative error code.
+ */
+int adf4382_set_cp_i(struct adf4382_dev *dev, int32_t reg_val)
+{
+	dev->cp_i = (uint8_t)reg_val;
+
+	if(reg_val > ADF4382_CPI_VAL_MAX)
+		dev->cp_i = ADF4382_CPI_VAL_MAX;
+
+	return adf4382_set_freq(dev);
+}
+
+/**
+ * @brief Gets the charge pump value from the register. The value will be
+ * between 0 and 15 on 8 bits. For more information please consult the
+ * Datasheet.
+ * @param dev 		- The device structure.
+ * @param reg_val 	- The read charge pump register value.
+ * @return    		- 0 in case of success or negative error code.
+ */
+int adf4382_get_cp_i(struct adf4382_dev *dev, int32_t *reg_val)
+{
+	uint8_t tmp;
+	uint8_t ret;
+
+	ret = adf4382_spi_read(dev, 0x1F, &tmp);
+	if (ret)
+		return ret;
+
+	dev->cp_i = no_os_field_get(ADF4382_CP_I_MSK, tmp);
+	*reg_val = dev->cp_i;
+
+	return 0;
+}
+
+/**
+ * @brief Set the bleed word, which represents the value of the bleed current
+ * written to the register space.
+ * @param dev 		- The device structure.
+ * @param word	 	- The bleed current register value.
+ * @return    		- 0 in case of success or negative error code.
+ */
+int adf4382_set_bleed_word(struct adf4382_dev *dev, int32_t word)
+{
+
+	dev->bleed_word = (uint16_t)word;
+
+	if(word > ADF4382_BLEED_WORD_MAX)
+		dev->bleed_word = ADF4382_BLEED_WORD_MAX;
+
+	return adf4382_set_freq(dev);
+}
+
+/**
+ * @brief Gets the value of the set bleed word.
+ * @param dev 		- The device structure.
+ * @param word	 	- The read bleed current register value.
+ * @return    		- Result of the reading procedure, error code otherwise.
+ */
+int adf4382_get_bleed_word(struct adf4382_dev *dev, int32_t *word)
+{
+	uint8_t upper;
+	uint8_t lower;
+	int ret;
+
+	ret = adf4382_spi_read(dev, 0x1E, &upper);
+	if (ret)
+		return ret;
+
+	upper &= (ADF4382_COARSE_BLEED_MSK | ADF4382_FINE_BLEED_MSB_MSK);
+
+	ret = adf4382_spi_read(dev, 0x1D, &lower);
+	if(ret)
+		return ret;
+
+	dev->bleed_word = (upper << 8) | lower;
+	*word = dev->bleed_word;
+
+	return 0;
+}
+
+/**
+ * @brief Set the desired output frequency and reset everything over to maximum
+ * supported value of 22GHz (21GHz for ADF4382A) to the max. value and
+ * everything under the minimum supported value of 687.5MHz (2.875GHz for
+ * ADF4382A) to the min. value.
+ * @param dev 		- The device structure.
+ * @param val	 	- The desired output frequency in Hz.
+ * @return    		- 0 in case of success or negative error code.
+ */
+int adf4382_set_rfout(struct adf4382_dev *dev, uint64_t val)
+{
+	dev->freq= val;
+
+	if(val > dev->freq_max)
+		dev->freq = dev->freq_max;
+
+	if(val < dev->freq_min)
+		dev->freq = dev->freq_min;
+
+	return adf4382_set_freq(dev);
+}
+
+/**
+ * @brief Gets the user proposed output frequency
+ * @param dev 		- The device structure.
+ * @param val		- The set value of the output frequency in Hz.
+ * @return    		- Nothing
+ */
+void adf4382_get_rfout(struct adf4382_dev *dev, uint64_t *val)
+{
+	*val = dev->freq;
```

Comment:
this should be obtained from hardware.

Reply:
The value is not readily available from the registers and would need to be recalculated from multiple registers. this would require several multiplication and division operations and would loose precision along the way.